ddtrace/internal/processor/trace.py

```diff
@@ -159,6 +159,16 @@ def on_span_finish(self, span):
                     except Exception:
                         log.error("error applying processor %r", tp, exc_info=True)
 
+                # mark top level span
+                if spans:
+                    chunk_root = spans[0]
+                    if chunk_root is chunk_root._local_root:
+                        chunk_root.set_metric("_dd.top_level", 1)
+                    elif chunk_root._parent and chunk_root._parent.service != chunk_root.service:
+                        chunk_root.set_metric("_dd.top_level", 1)
+                    elif should_partial_flush:
+                        chunk_root.set_metric("_dd.top_level", 0)
```

Comment:
Should I move this logic to a trace processor (ex. TraceTagsProcessor)? It seems out of place here

Reply:
or it could be it's own trace processor `TopLevelTraceProcessor` or something. then we'll get better separation of concerns.

what do you think?

Reply:
besides that, I think this logic makes sense. Originally I was thinking we'd want this in `Tracer.start_span`, or even an `on_start_span` hook, but this approach is likely more performant since we only need to look at and consider the spans that the trace agent currently messes up when partial flushing is enabled.

Reply:
Yes a `TraceProcessor` would be preferred here. We'd lose the information about whether it's a partial flush or not but I think we can assume that if the first span in the chunk is not the local root then we are partial flushing anyway (there's no other case when this happens).